src/treq/test/test_treq_integration.py

```diff
@@ -241,6 +243,156 @@ def test_failed_basic_auth(self):
         self.assertEqual(response.code, 401)
         yield print_response(response)
 
+    @inlineCallbacks
+    def test_digest_auth(self):
+        """
+            Test successful Digest authentication
+        :return:
+        """
+        response = yield self.get('/digest-auth/auth/treq/treq',
+                                  auth=HTTPDigestAuth('treq', 'treq'))
+        self.assertEqual(response.code, 200)
+        yield print_response(response)
+        json = yield treq.json_content(response)
+        self.assertTrue(json['authenticated'])
+        self.assertEqual(json['user'], 'treq')
+
+    @inlineCallbacks
+    def test_digest_auth_multi_qop(self):
+        """
+            Test successful Digest authentication with multiple qop types
+        :return:
+        """
+        response = yield self.get('/digest-auth/undefined/treq/treq',
+                                  auth=HTTPDigestAuth('treq', 'treq'))
+        self.assertEqual(response.code, 200)
+        yield print_response(response)
+        json = yield treq.json_content(response)
+        self.assertTrue(json['authenticated'])
+        self.assertEqual(json['user'], 'treq')
+
+    @inlineCallbacks
+    def test_digest_auth_multiple_calls(self):
+        """
+            Test proper Digest authentication credentials caching
+        """
+
+        # A mutable holder for call counter
+        agent_request_call_storage = {
+            'c': 0,
+            'i': []
+        }
+
+        # Original Agent request call
+        agent_request_orig = Agent.request
+
+        def agent_request_patched(*args, **kwargs):
+            """
+                Patched Agent.request function,
+                that inscreaces call count on every HTTP request
+                and appends
+            """
+            response_deferred = agent_request_orig(*args, **kwargs)
+            agent_request_call_storage['c'] += 1
+            agent_request_call_storage['i'].append((args, kwargs))
+            return response_deferred
+
+        patcher = MonkeyPatcher(
```

Comment:
Can't `self.patch` accomplish this?

https://docs.twistedmatrix.com/en/stable/api/twisted.trial.unittest.SynchronousTestCase.html#patch

Reply:
> Can't `self.patch` accomplish this?

Yes, however I had to rename one of the `with_baseurl` functions to not conflict:
```diff
-    patch = with_baseurl(treq.patch)
+    patch_req = with_baseurl(treq.patch)
```